File: src/src/main/resources/mmlspark/_ImageFeaturizer.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *
from mmlspark.TypeConversionUtils import generateTypeConverter, complexTypeConverter

@inherit_doc
class _ImageFeaturizer(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """


    Args:

        cntkModel (object): The internal CNTK model used in the featurizer
        cutOutputLayers (int): The number of layers to cut off the end of the network, 0 leaves the network intact, 1 removes the output layer, etc (default: 1)
        dropNa (bool): Whether to drop na values before mapping (default: true)
        inputCol (str): The name of the input column
        layerNames (list): Array with valid CNTK nodes to choose from, the first entries of this array should be closer to the output node
        outputCol (str): The name of the output column (default: [self.uid]_output)
    """

    @keyword_only
    def __init__(self, cntkModel=None, cutOutputLayers=1, dropNa=True, inputCol=None, layerNames=None, outputCol=None):
        super(_ImageFeaturizer, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.ImageFeaturizer")
        self._cache = {}
        self.cntkModel = Param(self, "cntkModel", "cntkModel: The internal CNTK model used in the featurizer", generateTypeConverter("cntkModel", self._cache, complexTypeConverter))
        self.cutOutputLayers = Param(self, "cutOutputLayers", "cutOutputLayers: The number of layers to cut off the end of the network, 0 leaves the network intact, 1 removes the output layer, etc (default: 1)")
        self._setDefault(cutOutputLayers=1)
        self.dropNa = Param(self, "dropNa", "dropNa: Whether to drop na values before mapping (default: true)")
        self._setDefault(dropNa=True)
        self.inputCol = Param(self, "inputCol", "inputCol: The name of the input column")
        self.layerNames = Param(self, "layerNames", "layerNames: Array with valid CNTK nodes to choose from, the first entries of this array should be closer to the output node")
        self.outputCol = Param(self, "outputCol", "outputCol: The name of the output column (default: [self.uid]_output)")
        self._setDefault(outputCol=self.uid + "_output")
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, cntkModel=None, cutOutputLayers=1, dropNa=True, inputCol=None, layerNames=None, outputCol=None):
        """
        Set the (keyword only) parameters

        Args:

            cntkModel (object): The internal CNTK model used in the featurizer
            cutOutputLayers (int): The number of layers to cut off the end of the network, 0 leaves the network intact, 1 removes the output layer, etc (default: 1)
            dropNa (bool): Whether to drop na values before mapping (default: true)
            inputCol (str): The name of the input column
            layerNames (list): Array with valid CNTK nodes to choose from, the first entries of this array should be closer to the output node
            outputCol (str): The name of the output column (default: [self.uid]_output)
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setCntkModel(self, value):
        """

        Args:

            cntkModel (object): The internal CNTK model used in the featurizer

        """
        self._set(cntkModel=value)
        return self


    def getCntkModel(self):
        """

        Returns:

            object: The internal CNTK model used in the featurizer
        """
        return self._cache.get("cntkModel", None)


    def setCutOutputLayers(self, value):
        """

        Args:

            cutOutputLayers (int): The number of layers to cut off the end of the network, 0 leaves the network intact, 1 removes the output layer, etc (default: 1)

        """
        self._set(cutOutputLayers=value)
        return self


    def getCutOutputLayers(self):
        """

        Returns:

            int: The number of layers to cut off the end of the network, 0 leaves the network intact, 1 removes the output layer, etc (default: 1)
        """
        return self.getOrDefault(self.cutOutputLayers)


    def setDropNa(self, value):
        """

        Args:

            dropNa (bool): Whether to drop na values before mapping (default: true)

        """
        self._set(dropNa=value)
        return self


    def getDropNa(self):
        """

        Returns:

            bool: Whether to drop na values before mapping (default: true)
        """
        return self.getOrDefault(self.dropNa)


    def setInputCol(self, value):
        """

        Args:

            inputCol (str): The name of the input column

        """
        self._set(inputCol=value)
        return self


    def getInputCol(self):
        """

        Returns:

            str: The name of the input column
        """
        return self.getOrDefault(self.inputCol)


    def setLayerNames(self, value):
        """

        Args:

            layerNames (list): Array with valid CNTK nodes to choose from, the first entries of this array should be closer to the output node

        """
        self._set(layerNames=value)
        return self


    def getLayerNames(self):
        """

        Returns:

            list: Array with valid CNTK nodes to choose from, the first entries of this array should be closer to the output node
        """
        return self.getOrDefault(self.layerNames)


    def setOutputCol(self, value):
        """

        Args:

            outputCol (str): The name of the output column (default: [self.uid]_output)

        """
        self._set(outputCol=value)
        return self


    def getOutputCol(self):
        """

        Returns:

            str: The name of the output column (default: [self.uid]_output)
        """
        return self.getOrDefault(self.outputCol)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.ImageFeaturizer"

    @staticmethod
    def _from_java(java_stage):
        module_name=_ImageFeaturizer.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".ImageFeaturizer"
        return from_java(java_stage, module_name)

File: src/main/scala/com/microsoft/ml/spark/image/ImageFeaturizer.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark.image

import com.microsoft.CNTK.CNTKExtensions._
import com.microsoft.CNTK.{SerializableFunction => CNTKFunction}
import com.microsoft.ml.spark.cntk.CNTKModel
import com.microsoft.ml.spark.core.contracts.{HasInputCol, HasOutputCol, Wrappable}
import com.microsoft.ml.spark.core.env.InternalWrapper
import com.microsoft.ml.spark.core.schema.{DatasetExtensions, ImageSchemaUtils}
import com.microsoft.ml.spark.downloader.ModelSchema
import org.apache.spark.ml.{ComplexParamsReadable, ComplexParamsWritable, Transformer}
import org.apache.spark.ml.linalg.SQLDataTypes.VectorType
import org.apache.spark.ml.param._
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.sql.types.{BinaryType, StructType}
import org.apache.spark.sql.{DataFrame, Dataset}

object ImageFeaturizer extends ComplexParamsReadable[ImageFeaturizer]

/** The <code>ImageFeaturizer</code> relies on a CNTK model to do the featurization, one can set
  * this model using the <code>modelLocation</code> parameter. To map the nodes of the CNTK model
  * onto the standard "layers" structure of a feed forward neural net, one needs to supply a list of
  * node names that range from the output node, back towards the input node of the CNTK Function.
  * This list does not need to be exhaustive, and is provided to you if you use a model downloaded
  * from the <code>ModelDownloader</code>, one can find this layer list in the schema of the
  * downloaded model.
  *
  * The <code>ImageFeaturizer</code> takes an input column of images (the type returned by the
  * <code>ImageReader</code>), and automatically resizes them to fit the CMTKModel's inputs.  It
  * then feeds them through a pre-trained CNTK model.  One can truncate the model using the <code>
  * cutOutputLayers </code> parameter that determines how many layers to truncate from the output of
  * the network.  For example, layer=0 means that no layers are removed, layer=2 means that the
  * image featurizer returns the activations of the layer that is two layers from the output layer.
  *
  * @param uid the uid of the image transformer
  */
@InternalWrapper
class ImageFeaturizer(val uid: String) extends Transformer with HasInputCol with HasOutputCol
  with Wrappable with ComplexParamsWritable {
  def this() = this(Identifiable.randomUID("ImageFeaturizer"))

  // Parameters related to the inner model

  val cntkModel: TransformerParam =
    new TransformerParam(
      this,
      "cntkModel", "The internal CNTK model used in the featurizer",
      { t => t.isInstanceOf[CNTKModel] })

  /** @group setParam */
  def setCntkModel(value: CNTKModel): this.type = set(cntkModel, value)

  val emptyCntkModel = new CNTKModel()

  /** @group getParam */
  def getCntkModel: CNTKModel =
    if (isDefined(cntkModel)) $(cntkModel).asInstanceOf[CNTKModel] else emptyCntkModel

  /** @group setParam */
  def setMiniBatchSize(value: Int): this.type = set(cntkModel, getCntkModel.setMiniBatchSize(value))

  /** @group getParam */
  def getMiniBatchSize: Int = getCntkModel.getMiniBatchSize

  /** @group setParam */
  def setInputNode(value: Int): this.type = set(cntkModel, getCntkModel.setInputNodeIndex(value))

  /** @group getParam */
  def getInputNode: Int = getCntkModel.getInputNodeIndex

  def setModelLocation(path: String): this.type = {
    set(cntkModel, getCntkModel.setModelLocation(path))
  }

  def setModel(modelSchema: ModelSchema): this.type = {
    setLayerNames(modelSchema.layerNames)
      .setInputNode(modelSchema.inputNode)
      .setModelLocation(modelSchema.uri.toString)
  }

  /** @group setParam */
  def setModel(model: CNTKFunction): this.type = set(cntkModel, getCntkModel.setModel(model))

  /** @group getParam */
  def getModel: CNTKFunction = getCntkModel.getModel

  // Parameters for just the ImageFeaturizer

  /** The number of layer to cut off the end of the network; 0 leaves the network intact, 1 removes
    * the output layer, etc.
    *
    * @group param
    */
  val cutOutputLayers: IntParam = new IntParam(this, "cutOutputLayers", "The number of layers to cut " +
    "off the end of the network, 0 leaves the network intact," +
    " 1 removes the output layer, etc", ParamValidators.gtEq(0))

  /** @group setParam */
  def setCutOutputLayers(value: Int): this.type = set(cutOutputLayers, value)

  /** @group getParam */
  def getCutOutputLayers: Int = $(cutOutputLayers)

  //TODO make nulls pass through
  val dropNa: BooleanParam =
    new BooleanParam(this, "dropNa", "Whether to drop na values before mapping")

  /** @group setParam */
  def setDropNa(value: Boolean): this.type = set(dropNa, value)

  /** @group getParam */
  def getDropNa: Boolean = $(dropNa)

  /** Array with valid CNTK nodes to choose from; the first entries of this array should be closer
    * to the output node.
    *
    * @group param
    */
  val layerNames: StringArrayParam = new StringArrayParam(this, "layerNames",
    "Array with valid CNTK nodes to choose from, the first entries of" +
      " this array should be closer to the output node")

  /** @group setParam */
  def setLayerNames(value: Array[String]): this.type = set(layerNames, value)

  /** @group getParam */
  def getLayerNames: Array[String] = $(layerNames)

  setDefault(cutOutputLayers -> 1, outputCol -> (uid + "_output"), dropNa->true)

  override def transform(dataset: Dataset[_]): DataFrame = {
    val resizedCol = DatasetExtensions.findUnusedColumnName("resized")(dataset.columns.toSet)

    val cntkModel = getCntkModel
      .setOutputNode(getLayerNames.apply(getCutOutputLayers))
      .setInputCol(resizedCol)
      .setOutputCol(getOutputCol)

    val requiredSize = cntkModel.getModel.getArguments.get(0).getShape.getDimensions

    val inputSchema = dataset.schema(getInputCol).dataType

    val unrolledDF = if (ImageSchemaUtils.isImage(inputSchema)) {
      val prepare = new ResizeImageTransformer()
        .setInputCol(getInputCol)
        .setWidth(requiredSize(0).toInt)
        .setHeight(requiredSize(1).toInt)
        .setNChannels(3)

      val unroll = new UnrollImage()
        .setInputCol(prepare.getOutputCol)
        .setOutputCol(resizedCol)

      val resizedDF = prepare.transform(dataset)
      unroll.transform(resizedDF).drop(prepare.getOutputCol)
    } else if (inputSchema == BinaryType) {
      val unroll = new UnrollBinaryImage()
        .setInputCol(getInputCol)
        .setWidth(requiredSize(0).toInt)
        .setHeight(requiredSize(1).toInt)
        .setNChannels(3)
        .setOutputCol(resizedCol)
      unroll.transform(dataset)
    } else {
      throw new IllegalArgumentException(
        s"Input schema : $inputSchema needs to have image or binary type")
    }

    val dropped = if (getDropNa){
      unrolledDF.na.drop(List(resizedCol))
    }else{
      unrolledDF
    }
    cntkModel.transform(dropped).drop(resizedCol)

  }

  override def copy(extra: ParamMap): Transformer = defaultCopy(extra)

  /** Add the features column to the schema
    *
    * @param schema
    * @return schema with features column
    */
  override def transformSchema(schema: StructType): StructType = {
    schema.add(getOutputCol, VectorType)
  }

}
